Index: qemu-0.10.0/linux-user/syscall.c
===================================================================
--- qemu-0.10.0.orig/linux-user/syscall.c	2009-03-07 13:47:26.000000000 +0100
+++ qemu-0.10.0/linux-user/syscall.c	2009-03-07 13:47:26.000000000 +0100
@@ -46,8 +46,10 @@
 #include <sys/uio.h>
 #include <sys/poll.h>
 #include <sys/times.h>
+#include <sys/ipc.h>
 #include <sys/shm.h>
 #include <sys/sem.h>
+#include <sys/msg.h>
 #include <sys/statfs.h>
 #include <utime.h>
 #include <sys/sysinfo.h>
@@ -2296,7 +2298,8 @@
             ret = get_errno(shmctl(first, second, NULL));
             break;
         default:
-            goto unimplemented;
+            ret = get_errno(shmctl(first, second, (struct shmid_ds *) ptr));
+            break;
         }
         break;
     default:
