Index: qemu-0.10.0/hw/pc.c
===================================================================
--- qemu-0.10.0.orig/hw/pc.c	2009-03-07 13:47:27.000000000 +0100
+++ qemu-0.10.0/hw/pc.c	2009-03-07 13:47:27.000000000 +0100
@@ -938,7 +938,7 @@
     ferr_irq = i8259[13];
 
     if (pci_enabled) {
-        pci_bus = i440fx_init(&i440fx_state, i8259);
+        pci_bus = i440fx_init(&i440fx_state, i8259, ram_size);
         piix3_devfn = piix3_init(pci_bus, -1);
     } else {
         pci_bus = NULL;
Index: qemu-0.10.0/hw/pc.h
===================================================================
--- qemu-0.10.0.orig/hw/pc.h	2009-03-07 13:46:58.000000000 +0100
+++ qemu-0.10.0/hw/pc.h	2009-03-07 13:47:27.000000000 +0100
@@ -112,7 +112,7 @@
 int pcspk_audio_init(AudioState *, qemu_irq *pic);
 
 /* piix_pci.c */
-PCIBus *i440fx_init(PCIDevice **pi440fx_state, qemu_irq *pic);
+PCIBus *i440fx_init(PCIDevice **pi440fx_state, qemu_irq *pic, int ram_size);
 void i440fx_set_smm(PCIDevice *d, int val);
 int piix3_init(PCIBus *bus, int devfn);
 void i440fx_init_memory_mappings(PCIDevice *d);
Index: qemu-0.10.0/hw/piix_pci.c
===================================================================
--- qemu-0.10.0.orig/hw/piix_pci.c	2009-03-07 13:46:58.000000000 +0100
+++ qemu-0.10.0/hw/piix_pci.c	2009-03-07 13:47:27.000000000 +0100
@@ -169,7 +169,7 @@
     return 0;
 }
 
-PCIBus *i440fx_init(PCIDevice **pi440fx_state, qemu_irq *pic)
+PCIBus *i440fx_init(PCIDevice **pi440fx_state, qemu_irq *pic, int ram_size)
 {
     PCIBus *b;
     PCIDevice *d;
@@ -197,6 +197,10 @@
     d->config[0x08] = 0x02; // revision
     pci_config_set_class(d->config, PCI_CLASS_BRIDGE_HOST);
     d->config[0x0e] = 0x00; // header_type
+    ram_size = ram_size / 8 / 1024 / 1024;
+    if (ram_size > 255)
+        ram_size = 255;
+    d->config[0x57] = ram_size;
 
     d->config[0x72] = 0x02; /* SMRAM */
 
