Index: qemu-0.10.0/linux-user/arm/syscall_nr.h
===================================================================
--- qemu-0.10.0.orig/linux-user/arm/syscall_nr.h	2009-03-07 13:46:58.000000000 +0100
+++ qemu-0.10.0/linux-user/arm/syscall_nr.h	2009-03-07 13:47:27.000000000 +0100
@@ -256,9 +256,8 @@
 #define TARGET_NR_epoll_ctl			(251)
 #define TARGET_NR_epoll_wait			(252)
 #define TARGET_NR_remap_file_pages		(253)
-					/* 254 for set_thread_area */
-					/* 255 for get_thread_area */
-					/* 256 for set_tid_address */
+#define TARGET_NR_set_thread_area		(254)
+#define TARGET_NR_get_thread_area		(255)
 #define TARGET_NR_set_tid_address		256
 #define TARGET_NR_timer_create		257
 #define TARGET_NR_timer_settime		258
