Index: qemu-0.10.0/ppc.ld
===================================================================
--- qemu-0.10.0.orig/ppc.ld	2009-03-07 13:46:58.000000000 +0100
+++ qemu-0.10.0/ppc.ld	2009-03-07 13:47:28.000000000 +0100
@@ -8,6 +8,7 @@
 SECTIONS
 {
   /* Read-only sections, merged into text segment: */
+  PROVIDE (__executable_start = 0x60000000);
   . = 0x60000000 + SIZEOF_HEADERS;
   .interp     : { *(.interp) 	}
   .hash          : { *(.hash)		}
@@ -210,7 +211,7 @@
   .debug_aranges  0 : { *(.debug_aranges) }
   .debug_pubnames 0 : { *(.debug_pubnames) }
   /* DWARF 2 */
-  .debug_info     0 : { *(.debug_info) }
+  .debug_info     0 : { *(.debug_info .gnu.linkonce.wi.*) }
   .debug_abbrev   0 : { *(.debug_abbrev) }
   .debug_line     0 : { *(.debug_line) }
   .debug_frame    0 : { *(.debug_frame) }
